File: 2022/Day6.hs
#! /usr/bin/env -S runhaskell
import Data.List
import Data.List.Split
import Data.Char
import Data.Maybe (fromJust)

import Util

type Input = String
type Output = Int

parse :: String -> Maybe Input
parse = Just

solve :: Int -> Input -> Maybe Output
solve n = ((+n) <$>) . findIndex ((==) <$> nub <*> id) . map (take n) . tails

part1 :: Input -> Output
part1 = fromJust . solve 4

part2 :: Input -> Output
part2 = fromJust . solve 14

main :: IO()
main = do
    f <- readFile("6.in")
    let input = parse f
    case input of
      Nothing -> putStrLn "Unable to read input."
      Just i -> do
        putStr "Solution for part 1: "
        putStrLn . show $ part1 i
        putStr "Solution for part 2: "
        putStrLn . show $ part2 i

-- Solution part 1: 1598
-- Solution part 2: 2414

File: 2022/Day6.py
#! /usr/bin/env python
# pylint: disable=unused-wildcard-import
from util import *

def get_marker(inp, n):
    for i in range(len(inp)):
        if len(set(inp[i:i+n])) == n:
            return i + n
    return None

def p1(inp):
    return get_marker(inp, 4)

def p2(inp):
    return get_marker(inp, 14)

def main():
    inp = data()
    print(f"Solution for part 1:\n{p1(inp)}")
    print(f"Solution for part 2:\n{p2(inp)}")

if __name__ == "__main__":
    main()

# Solution part 1: 1598
# Solution part 2: 2414
